upgrade_analysis/odoo_patch/odoo/models.py

```diff
@@ -1,179 +1,21 @@
-                    info = data_list[0]['info']
-                    messages.append(dict(info, type='error', message=_(u"Unknown database error: '%s'", e)))
-                return
-            except Exception:
-                pass
-
-            errors = 0
-            # try again, this time record by record
-            for i, rec_data in enumerate(data_list, 1):
-                try:
-                    with cr.savepoint():
-                        rec = self._load_records([rec_data], mode == 'update')
-                        ids.append(rec.id)
-                except psycopg2.Warning as e:
-                    info = rec_data['info']
-                    messages.append(dict(info, type='warning', message=str(e)))
-                except psycopg2.Error as e:
-                    info = rec_data['info']
-                    messages.append(dict(info, type='error', **PGERROR_TO_OE[e.pgcode](self, fg, info, e)))
-                    # Failed to write, log to messages, rollback savepoint (to
-                    # avoid broken transaction) and keep going
-                    errors += 1
-                except Exception as e:
-                    _logger.debug("Error while loading record", exc_info=True)
-                    info = rec_data['info']
-                    message = (_(u'Unknown error during import:') + u' %s: %s' % (type(e), e))
-                    moreinfo = _('Resolve other errors first')
-                    messages.append(dict(info, type='error', message=message, moreinfo=moreinfo))
-                    # Failed for some reason, perhaps due to invalid data supplied,
-                    # rollback savepoint and keep going
-                    errors += 1
-                if errors >= 10 and (errors >= i / 10):
-                    messages.append({
-                        'type': 'warning',
-                        'message': _(u"Found more than 10 errors and more than one error per 10 records, interrupted to avoid showing too many errors.")
-                    })
-                    break
-
-        # make 'flush' available to the methods below, in the case where XMLID
-        # resolution fails, for instance
-        flush_self = self.with_context(import_flush=flush)
-
-        # TODO: break load's API instead of smuggling via context?
-        limit = self._context.get('_import_limit')
-        if limit is None:
-            limit = float('inf')
-        extracted = flush_self._extract_records(fields, data, log=messages.append, limit=limit)
-
-        converted = flush_self._convert_records(extracted, log=messages.append)
-
-        info = {'rows': {'to': -1}}
-        for id, xid, record, info in converted:
+from odoo import api, models
+from ..odoo_patch import OdooPatch
+from ....upgrade_analysis import upgrade_log
+
+
+class BaseModelPatch(OdooPatch):
+    target = models.BaseModel
+    method_names = ['_convert_records']
+
+    @api.model
+    def _convert_records(self, records, log=lambda a: None):
```

Comment:
Huge simplification ! from 179 lines to 21 !

"yeld function" seem easier to patch. Very clean.